Feature: Added support for File Parts to enable complete multimodal inputs

shaunak99 opened the pull request:
This adds support for the LLM to support File Parts. Gemini and OpenAI (gpt-4o-audio-preview) support sending audio files. Gemini, OpenAI & Anthropic support sending PDFs natively (processed as images, page by page). 


<img width="1508" height="853" alt="Screenshot 2025-07-24 at 2 46 54 AM" src="https://github.com/user-attachments/assets/b1d5262c-14c3-40a7-a7d5-90bff21bb311" />

Also updated WebUI to support the same.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added support for sending and receiving file attachments (PDFs and audio files) in chat conversations.
  * Enabled audio recording directly within the chat input area.

* **Enhancements**
  * Chat messages now visually display and allow playback of audio files and show attached PDF files.
  * Updated chat input and message list UI to support and preview file attachments.
  * Expanded API documentation to detail supported file types and file upload capabilities.
  * Added model-aware filtering and validation for file attachments based on provider and model capabilities.
  * Improved message formatting to handle file content across multiple LLM providers.
  * Enhanced type safety and tooling around file data in chat and AI message handling.
  * Added comprehensive input validation for text, image, and file data with user feedback on unsupported inputs.
  * Introduced event signaling for unsupported input scenarios in LLM services.

* **Bug Fixes**
  * Adjusted tests to align with updated method signatures for file support.

* **Documentation**
  * Updated API and type documentation to reflect new file attachment features and supported formats.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

shaunak99 commented:
Thanks for the detailed feedback! Very insightful for understanding why and how to go about the implementation. 

I've tried to address most of it while reducing as much of the code duplication, and type fixes. Tested it with gemini-2.5-flash.

Check it out and let me know!

rahulkarajgikar commented on src/app/api/server.ts:
```diff
@@ -84,7 +84,50 @@ export async function initializeApi(agent: SaikiAgent, agentCardOverride?: Parti
+
+                // Security validation: MIME type allowlist
+                const allowedMimeTypes = [
+                    'application/pdf',
```
i see

we can add a separate issue for this probably since it would require some special handling 

i guess txt and md would be easy but doc formats would have challenges

rahulkarajgikar replied:
added #247

rahulkarajgikar commented on src/app/api/server.ts:
```diff
@@ -84,7 +84,73 @@ export async function initializeApi(agent: SaikiAgent, agentCardOverride?: Parti
+                }
+
+                // Security validation: MIME type allowlist
+                const allowedMimeTypes = [
```
move to ALLOWED_MIME_TYPES constant (coderabbit has similar cmt)

rahulkarajgikar commented on src/core/ai/llm/registry.ts:
```diff
@@ -280,6 +347,202 @@ export function isValidProvider(provider: string): boolean {
+ * @param provider The name of the provider.
+ * @returns Array of supported file types for the provider, or empty array if provider not found.
+ */
+export function getSupportedFileTypes(provider: string): SupportedFileType[] {
```
i don't think we need provider level functions or defaults

let's keep it at model level only, generalizing across a provider can be unreliable, some models for that provider may support, some models may not.

can remove all the functions that accept only provider and not model

rahulkarajgikar commented on src/app/api/server.ts:
```diff
@@ -124,6 +124,29 @@ export async function initializeApi(agent: SaikiAgent, agentCardOverride?: Parti
+                    '@core/ai/llm/validation.js'
+                );
+
+                const validation = validateFileForLLM(
```
we can make a more generic function validateInputForLLM - that accepts the current raw inputs and LLM(image, filedata)

that can check for both image supported and file supported

we also don't need the checks from lines 93->126, this is duplicate checks which we can handle once in the validation function

overall api code should be pretty short:

1. extract inputs
2. get current llm config for session
3. validate entire input + llm combination (can delegate to the validate function)
4. return response

similar for APIs below

rahulkarajgikar commented on src/app/webui/components/MessageList.tsx:
```diff
@@ -2,7 +2,16 @@
 
 import React, { useEffect, useRef, useState } from 'react';
```
not reviewing this file, just trusting that this works

rahulkarajgikar commented on src/app/webui/components/hooks/useChat.ts:
```diff
@@ -12,15 +12,61 @@ export interface ImagePart {
+
+export function isFilePart(part: unknown): part is FilePart {
+    return typeof part === 'object' && part !== null && (part as any).type === 'file';
+}
```
yeah remove any type

rahulkarajgikar commented on src/core/ai/llm/messages/formatters/openai.ts:
```diff
@@ -18,9 +23,26 @@ export class OpenAIMessageFormatter implements IMessageFormatter {
+    format(
+        history: Readonly<InternalMessage[]>,
+        systemPrompt: string | null,
+        context?: any
```
`any` type

rahulkarajgikar commented on src/core/ai/llm/messages/formatters/types.ts:
```diff
@@ -1,5 +1,20 @@
+ */
+export interface FormatterContext {
+    /** MCP manager for tool handling */
+    mcpManager?: any; // TODO: Replace with proper MCPManager type when available
```
we have the mcpManager type

rahulkarajgikar commented on src/core/ai/llm/messages/formatters/types.ts:
```diff
@@ -1,5 +1,20 @@
+    mcpManager?: any; // TODO: Replace with proper MCPManager type when available
+
+    /** LLM provider name (e.g., 'google.generative-ai', 'openai') */
+    llmProvider?: string;
```
can just stick to provider/model convention (no need for llm prefix - more consistent)

rahulkarajgikar commented on src/core/ai/llm/messages/formatters/vercel.ts:
```diff
@@ -23,9 +30,39 @@ export class VercelMessageFormatter implements IMessageFormatter {
+    format(
+        history: Readonly<InternalMessage[]>,
+        systemPrompt: string | null,
+        context?: any
```
no any type

rahulkarajgikar commented on src/core/ai/llm/messages/utils.ts:
```diff
@@ -108,3 +125,112 @@ export function getImageData(imagePart: {
+    return '';
+}
+
+import { validateFileForLLM } from '../validation.js';
```
+1

rahulkarajgikar commented on src/core/ai/llm/messages/formatters/vercel.ts:
```diff
@@ -23,9 +30,39 @@ export class VercelMessageFormatter implements IMessageFormatter {
+                filteredHistory = filterMessagesByCapabilities([...history], 'groq');
+            }
+        } catch (error) {
+            console.warn('Failed to apply capability filtering, using original history:', error);
```
+1

rahulkarajgikar commented on src/core/ai/llm/messages/manager.ts:
```diff
@@ -525,7 +525,7 @@ export class ContextManager {
             // Use pre-computed system prompt if provided
             const prompt = systemPrompt ?? (await this.getSystemPrompt(context));
-            return this.formatter.format([...messageHistory], prompt);
+            return this.formatter.format([...messageHistory], prompt, context);
```
trusting that this works cuz we don't have tests, let's just validate this manually

rahulkarajgikar commented on src/core/ai/llm/messages/utils.ts:
```diff
@@ -149,3 +149,88 @@ export function getFileData(filePart: {
+    }
+}
+
+/**
+ * @deprecated Use filterMessagesByLLMCapabilities instead
+ * Filters message content based on LLM provider capabilities.
+ */
+export function filterMessagesByCapabilities(
+    messages: InternalMessage[],
+    provider: string
+): InternalMessage[] {
+    return filterMessagesByLLMCapabilities(messages, { provider });
+}
+
+/**
+ * @deprecated Use filterMessagesByLLMCapabilities instead
+ * Filters message content based on specific model capabilities.
+ */
+export function filterMessagesByModelCapabilities(
+    messages: InternalMessage[],
+    provider: string,
+    model: string
+): InternalMessage[] {
+    return filterMessagesByLLMCapabilities(messages, { provider, model });
+}
```
delete

to avoid this kind of generation, add the prompt 'don't worry about backward compatibility' for PR refactors

rahulkarajgikar commented on src/core/ai/llm/registry.ts:
```diff
@@ -37,48 +41,108 @@ export type LLMProvider =
         ],
         supportedRouters: ['vercel', 'in-built'],
         baseURLSupport: 'none',
+        supportedFileTypes: ['pdf'], // Default for OpenAI models
```
let's not keep provider level defaults

we will always know what the current model is at any point in time, so we can have exact checks based on what the llm is

can also enforce that every entry for every LLM in this registry has the `supportedFileTypes` entry

rahulkarajgikar commented on src/core/ai/llm/registry.ts:
```diff
@@ -280,6 +347,202 @@ export function isValidProvider(provider: string): boolean {
+ * @param provider The provider name to normalize.
+ * @returns The normalized provider name that matches our registry.
+ */
+export function normalizeProviderName(provider: string): string {
```
don't think we need this, we check inputs way beforehand at the yaml schema, we wouldn't have inputs like `gemini` as the provider at this stage so normalization isn't necessary

rahulkarajgikar commented on src/core/ai/llm/registry.ts:
```diff
@@ -280,6 +347,202 @@ export function isValidProvider(provider: string): boolean {
+ * @param fileType The file type to check support for.
+ * @returns True if the provider supports the file type, false otherwise.
+ */
+export function supportsFileType(provider: string, fileType: SupportedFileType): boolean {
```
let's get rid of all functions that accept only provider and not model, we can just keep the ones that have provider + model combo

similar for tests

a good extension here would be to also add functions to check for supportsImageType

then we can have the validateInputForLLM function that checks the entire input (image? + file? + text + llm config ) from the APIs and returns supported/unsupported by calling the helper funcs

rahulkarajgikar commented on src/core/ai/llm/validation.ts:
```diff
@@ -0,0 +1,62 @@
+ * @param mimeType The MIME type of the file to validate
+ * @returns Validation result with support status and error details
+ */
+export function validateFileForLLM(
```
we can make this `validateInputForLLM` as mentioned earlier that accepts the entire API inputs

rahulkarajgikar commented on src/core/ai/llm/messages/utils.ts:
```diff
@@ -149,3 +149,88 @@ export function getFileData(filePart: {
+ * @param config The LLM configuration (provider and optional model)
+ * @returns Filtered messages with unsupported content removed
+ */
+export function filterMessagesByLLMCapabilities(
```
nice! this is a very useful helper function

let's add tests for this though, it's hard to validate edge cases manually

(combinations of audio, filetype for different models and validate the history that comes back)

rahulkarajgikar commented on src/core/ai/llm/registry.ts:
```diff
@@ -280,6 +347,202 @@ export function isValidProvider(provider: string): boolean {
+        'audio/ogg': 'audio',
+        'audio/m4a': 'audio',
+        'audio/aac': 'audio',
+    };
```
+1, a top level constant in the registry is the only place where we need this list

rahulkarajgikar commented on src/core/ai/session/chat-session.ts:
```diff
@@ -240,6 +240,42 @@ export class ChatSession {
         );
+
+        // Validate file format support if file data is provided
+        if (fileDataInput) {
```
with the validateInputForLLM function, this code would also become much better -> we'd just send the whole input + llm combination

we would get back some response, saying which part was supported, which was not, and can emit events accordingly

rahulkarajgikar commented on src/core/ai/systemPrompt/types.ts:
```diff
@@ -3,6 +3,9 @@ import { MCPManager } from '../../client/manager.js';
 export interface DynamicContributorContext {
     mcpManager: MCPManager;
+    // Model information for formatters - used for model-aware file filtering
+    llmProvider?: string;
```
keep `provider/model` convention without llm prefix

rahulkarajgikar reviewed (changes requested):
nice this is better! especially the `filterMessagesByLLMCapabilities` is a super useful utility

with the new suggested utility `validateInputForLLM` and this, we can cover most of the cases pretty well

these 2 methods + unit tests for them will cover most of the important cases

there is still quite a bit of code duplication which i've tried to address as much as possible 

some stuff we can still improve on:
-> add the registry metadata API for querying the llm registry from the web UI
-> then we can decide what user can upload directly at the UI level
-> this has the added benefit that we can add support for way more models in the web UI

can take this as follow up—as long as the web UI is handling unsupported input types gracefully for now that's good enough for this PR

shaunak99 commented:
Updated to `validateInputForLLM` and also added some tests for the input.

Also tried to do some more refactors and removed any types.

Model switching & filtering seems to work alright! 

<img width="1508" height="853" alt="Screenshot 2025-07-25 at 3 13 38 PM" src="https://github.com/user-attachments/assets/4027e538-a9b1-4b19-bf37-57cd53551841" />

rahulkarajgikar commented on src/app/api/server.ts:
```diff
@@ -430,94 +371,45 @@ export async function initializeApi(agent: SaikiAgent, agentCardOverride?: Parti
 
-                    const sessionId = data.sessionId as string | undefined;
                     const stream = data.stream === true; // Extract stream preference, default to false
                     if (imageDataInput) logger.info('Image data included in message.');
```
nit: move logs in line 414-416 before validation, similar in other apis also
flow would then be -> log the inputs, validate the inputs, return the response

currently if validation fails earlier we wouldn't see logs which could be harder to debug

rahulkarajgikar commented on src/core/ai/llm/messages/formatters/vercel.ts:
```diff
@@ -56,10 +55,12 @@ export class VercelMessageFormatter implements IMessageFormatter {
                 );
-                filteredHistory = filterMessagesByCapabilities([...history], 'groq');
+                filteredHistory = filterMessagesByLLMCapabilities([...history], {
+                    provider: 'groq',
```
why are we using groq?

there shouldn't be an else case here, context.provider and context.model have to be mandatory, otherwise we should throw an error. defaulting to other providers would cause weird behaviours

we should make provider and model non-optional in `FormatterContext` type and any other new types introduced that include provider and model. this would be like a pre-condition for any executions

rahulkarajgikar commented on src/core/ai/llm/messages/formatters/vercel.ts:
```diff
@@ -187,10 +190,10 @@ export class VercelMessageFormatter implements IMessageFormatter {
                                         type: 'image',
-                                        image: img.data,
-                                        mimeType: img.mimeType,
+                                        image: (img as any).data,
+                                        mimeType: (img as any).mimeType,
```
nit: llm cursed any kink is back. we might need to use/create an ImageData type for this

rahulkarajgikar commented on src/core/ai/llm/messages/utils.test.ts:
```diff
@@ -0,0 +1,322 @@
+        ]);
+    });
+
+    test('should add placeholder text without model name when no model specified', () => {
```
should fail when no model specified, make both provider and model mandatory

rahulkarajgikar commented on src/core/ai/llm/messages/utils.ts:
```diff
@@ -184,7 +183,15 @@ export function filterMessagesByLLMCapabilities(
                 if (part.type === 'file' && part.mimeType) {
-                    const validation = validateFileForLLM(config, part.mimeType);
+                    // If no model is specified, reject all files
+                    if (!config.model) {
+                        return false;
+                    }
```
raise error instead, this is an unexpected case

rahulkarajgikar commented on src/core/ai/llm/messages/utils.ts:
```diff
@@ -212,25 +219,4 @@ export function filterMessagesByLLMCapabilities(
-): InternalMessage[] {
-    return filterMessagesByLLMCapabilities(messages, { provider, model });
-}
+// Removed deprecated filter functions - no backward compatibility needed
```
stale line

rahulkarajgikar commented on src/core/ai/llm/validation.ts:
```diff
@@ -1,62 +1,175 @@
 
 /**
- * Creates a standardized error response for file validation failures.
+ * Validates image input (placeholder for future image validation logic).
+ * @param _imageData The image data to validate
+ * @param _config The LLM configuration
+ * @returns Image validation result
+ */
+function validateImageInput(
+    _imageData: ImageData,
+    _config: ValidationLLMConfig
+): NonNullable<InputValidationResult['imageValidation']> {
+    // For now, assume images are supported (existing behavior)
+    // This can be expanded later with proper image capability validation
+    return {
+        isSupported: true,
+    };
+}
```
let's add an issue for making this change

should be pretty straightforward change, can get community to pick it up

1. update registry to add supportsImage? for each LLM
2. add a function in registry.ts to check the registry
3. update this `validateImageInput` function
4. add tests

rahulkarajgikar commented on src/core/ai/session/chat-session.ts:
```diff
@@ -241,40 +242,7 @@ export class ChatSession {
-                `File validation passed for ${validation.fileType} file in session ${this.id}`
-            );
-        }
+        // Note: Input validation is now handled at the API layer before reaching this method
```
we still need to validate again here, because not all flows will go through API layer (SDK users)

we can re-use the validateInputForLLM.. func here and emit unsupported input events  in case of the validation errors. need to update event bus also to add this as a known event

rahulkarajgikar commented on src/core/ai/llm/validation.ts:
```diff
@@ -1,62 +1,175 @@
-    config: FileValidationConfig,
-    mimeType: string
-): FileValidationResult {
+export function validateInputForLLM(
```
`filterMessages` func has good coverage but we need some tests for this func as well as it's a critical point

rahulkarajgikar reviewed (changes requested):
changes look mostly good! few minor things still to be addressed, then we should be good

shaunak99 commented:
Addressed most of the comments. Will create the issue for image validation once we merge this.

rahulkarajgikar commented on src/core/ai/llm/messages/manager.ts:
```diff
@@ -525,7 +525,17 @@ export class ContextManager {
-            return this.formatter.format([...messageHistory], prompt, context);
+
+            // Map DynamicContributorContext to FormatterContext
+            const formatterContext: FormatterContext | undefined = context.llmProvider
```
nit: we should rechange dynamiccontributor to use provider/model convention. marking for my reference

rahulkarajgikar commented on src/core/ai/llm/validation.test.ts:
```diff
@@ -0,0 +1,417 @@
+}));
+
+describe('validateInputForLLM', () => {
+    const mockValidateModelFileSupport = vi.mocked(registry.validateModelFileSupport);
+    const mockGetAllowedMimeTypes = vi.mocked(registry.getAllowedMimeTypes);
+
+    beforeEach(() => {
+        vi.clearAllMocks();
+        // Default mock implementations
+        mockGetAllowedMimeTypes.mockReturnValue(['application/pdf', 'audio/mp3', 'audio/wav']);
+        mockValidateModelFileSupport.mockReturnValue({
+            isSupported: true,
+        });
```
@coderabbitai does it make sense to mock these functions - my thought was that we shoudl just use the direct registry functions and only mock for edge-cases where we don't have valid examples, wdyt? mocking these doesn't seem helpful to me (we test mock behaviour rather than real behaviour)

rahulkarajgikar replied:
ok i will handle this in https://github.com/truffle-ai/saiki/pull/249/files, can you suggest the tests

rahulkarajgikar reviewed (approved):
validation.test.ts might need some changes, waiting on coderabbit comments
